packages/flutter/lib/src/widgets/selectable_region.dart

```diff
@@ -3016,3 +3155,266 @@ typedef SelectableRegionContextMenuBuilder = Widget Function(
+/// an additional [SelectionListener] under each one.
+///
+/// {@tool dartpad}
+/// This example shows how to use [SelectionListener] to listen to selection changes
+/// under a [SelectionArea] or [SelectableRegion].
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.1.dart **
+/// {@end-tool}
+///
+/// {@tool dartpad}
+/// This example shows how to color the active selection red
+/// under a [SelectionArea] or [SelectableRegion].
+///
+/// ** See code in examples/api/lib/material/selection_area/selection_area.2.dart **
+/// {@end-tool}
+///
+/// See also:
+///
+///   * [SelectableRegion], which provides an overview of the selection system.
+class SelectionListener extends StatefulWidget {
+  /// Create a new [SelectionListener] widget.
+  const SelectionListener({
+    super.key,
+    required this.selectionNotifier,
+    required this.child,
+  });
+
+  /// Notifies listeners when the selection has changed.
+  final SelectionListenerNotifier selectionNotifier;
+
+  /// The child widget this selection listener applies to.
+  ///
+  /// {@macro flutter.widgets.ProxyWidget.child}
+  final Widget child;
+
+  @override
+  State<SelectionListener> createState() => _SelectionListenerState();
+}
+
+class _SelectionListenerState extends State<SelectionListener> {
+  late final _SelectionListenerDelegate _selectionDelegate = _SelectionListenerDelegate(selectionNotifier: widget.selectionNotifier);
+
+  @override
+  void didUpdateWidget(SelectionListener oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.selectionNotifier != widget.selectionNotifier) {
+      _selectionDelegate.setNotifier(widget.selectionNotifier);
+    }
+  }
+
+  @override
+  void dispose() {
+    _selectionDelegate.dispose();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return SelectionContainer(
+      delegate: _selectionDelegate,
+      child: widget.child,
+    );
+  }
+}
+
+final class _SelectionListenerDelegate extends StaticSelectionContainerDelegate implements SelectionDetails {
+  _SelectionListenerDelegate({
+    required SelectionListenerNotifier selectionNotifier,
+  }) : _selectionNotifier = selectionNotifier {
+    assert(
+      !_selectionNotifier.registered,
+      'The SelectionListenerNotifier provided is already attached to another SelectionListener. Try providing a new SelectionListenerNotifier.',
+    );
+    _selectionNotifier._selectionDelegate = this;
+  }
+
+  SelectionListenerNotifier _selectionNotifier;
+  void setNotifier(SelectionListenerNotifier newNotifier) {
+    _selectionNotifier = newNotifier;
+    _selectionNotifier._selectionDelegate = this;
+  }
+
+  SelectionGeometry? _previousSelectionGeometry;
+
+  @override
+  SelectionResult dispatchSelectionEvent(SelectionEvent event) {
+    final SelectionResult result = super.dispatchSelectionEvent(event);
+    if (_previousSelectionGeometry != value) {
+      _selectionNotifier.notifyListeners();
```

Comment:
should this only notifyListener when this `_SelectionListenerDelegate` notify its listener? this way you can catch selection change even if it is generated below this selectable

Reply:
Good point, I think that works out. The only difference I found when trying to hook in dispatchSelectionEvent vs notifyListeners is that hooking in notifyListeners dispatches a notification when the app is first run.

Reply:
I will suppress if the first one if possible, unless the initial one has a valid selection.